### PR TITLE
Fix testing for `pytest` 7.2 support

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

`pytest` 7.2 was just released. It deprecates several things that `astropy` was using for testing, including `setup` and `teardown` functions. This PR fixes all those deprecations by switching them over to `setup_method` and `teardown_method`. 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/coordinates/tests/test_celestial_transformations.py
Comment (pllim) [inline astropy/coordinates/tests/test_celestial_transformations.py:191]: Why not `setup_class`? Does this need to be redefined before every method test?

Same question for all other similar changes.

**Review by pllim — COMMENTED** [review]
Something else needs updating to get past the CI failure of this PR.

Thanks for tackling this!

File: astropy/coordinates/tests/test_celestial_transformations.py
Comment (WilliamJamieson) [reply astropy/coordinates/tests/test_celestial_transformations.py:191]: According to https://docs.pytest.org/en/7.2.x/deprecations.html#nose-deprecation `setup_method` `teardown_method` are the replacements for `setup` and `teardown`.

File: astropy/visualization/wcsaxes/tests/test_frame.py
Comment (pllim) [inline astropy/visualization/wcsaxes/tests/test_frame.py:75]: Are you sure this isn't needed for them weird hash image comparison that CircleCI picks up?

Comment (pllim) [reply astropy/visualization/wcsaxes/tests/test_frame.py:75]: I think this is intentional... #5496 cc @astrofrog

Comment (WilliamJamieson) [reply astropy/visualization/wcsaxes/tests/test_frame.py:75]: Per the pytest changelog:

> [#10196](https://github.com/pytest-dev/pytest/issues/10196):` [PytestReturnNotNoneWarning](https://docs.pytest.org/en/7.2.x/reference/reference.html#pytest.PytestReturnNotNoneWarning) is now a subclass of [PytestRemovedIn8Warning](https://docs.pytest.org/en/7.2.x/reference/reference.html#pytest.PytestRemovedIn8Warning): the plan is to make returning non-None from tests an error in the future.

Comment (pllim) [reply astropy/visualization/wcsaxes/tests/test_frame.py:75]: Looks like pytest isn't gonna back down, so we're blocked by https://github.com/matplotlib/pytest-mpl/issues/183 . I see two immediate actions we can take:

* Pin maxversion of pytest to `<7.2` to immediately unblock all the other PRs for feature freeze week. Turn this PR to draft until upstream can figure out the compatibility.
* Continue with this PR but use `pytest.mark.filterwarnings` to ignore this warning for the affected image tests. But this is also risky because if we forget and they turn warning into exception in a future release, we still have to pin maxversion then.

What to do?

File: astropy/tests/figures/helpers.py
Comment (pllim) [inline astropy/tests/figures/helpers.py:35]:
```suggestion
        @pytest.mark.skipif(not HAS_PYTEST_MPL, reason='pytest-mpl is required for the figure tests')
```

Comment (WilliamJamieson) [reply astropy/tests/figures/helpers.py:35]: Oops

**Review by pllim — APPROVED** [review]
The diff looks reasonable to me and skipping is fine, since we don't calculate coverage for remote data anyway. Thanks!

File: astropy/tests/figures/helpers.py
Comment (pllim) [inline astropy/tests/figures/helpers.py:35]: I thought about reverting this and use `-m "not mpl_image_compare"` as suggested in https://github.com/matplotlib/pytest-mpl/issues/183#issuecomment-1302733435 but I cannot figure out a sane way to define that in `tox.ini` because we have a few possible flag combos already and CircleCI can theoretically use any of those flags if it wants to.

https://github.com/astropy/astropy/blob/7e80b9fcfc43f1e720d3d3c8b6adcce8f3a47782/tox.ini#L119-L124

Maybe this is still the best option we have unless people have better ideas.

**Pull request merged**
